### DISCUSSION TITLE
[RFC] New Component: WaterMark 水印

### DISCUSSION THREAD

**Discussion by MadCcc** [ant-design/ant-design#38865] [RFCs - archive]
## Summary
新增 WaterMark 组件，给页面的某个区域加上水印。

## API

| 参数 | 说明 | 类型 | 默认值 | 
| --- | --- | --- | --- | 
| rotate | 水印绘制时，旋转的角度，单位 ° | number | -22 | 
| image | 图片源，建议导出 2 倍或 3 倍图，优先使用图片渲染水印 | `string` | - |
| zIndex | 追加的水印元素的 z-index | number | 9 |
| content | 水印文字内容 | `string` \| `string[]` | - | 
| font | 文字样式 | `{ color?: string, fontSize?: number, fontWeight?: 'normal' \| 'light' \| 'weight' \| number; fontFamily?: string; fontStyle?: 'none' \| 'normal' \| 'italic' \| 'oblique' }` | `{color?: 'rgba(0,0,0,.15)', fontSize?: 16, fontWeight?: 'normal'; fontFamily?: 'sans-serif'; fontStyle?: 'normal'}` |
| style | 容器层的样式 | React.CSSProperties | - | 
| className | 容器层的类名 | string | - | 
| gap | 水印之间的间距 | `[number, number]` | `[212, 222]` | 
| offset | 水印在 canvas 画布上绘制的偏移量, 正常情况下，水印绘制在中间位置, 即 `offset = gap / 2` | `[number, number]` | `[gap[0] / 2, gap[1] / 2]` | 

> 文字样式相关 props 都收到 `font` 中。

## Basic Example

https://procomponents.ant.design/components/water-mark

## Detailed Design
迁移 procomponents 下的 WaterMark 组件：https://github.com/ant-design/pro-components/tree/master/packages/layout/src/components/WaterMark
需要做的迁移工作包括但不限于：
1. style 改用 `genComponentStyleHook`，减少 hashId 的使用，token 调整；
2. 内部逻辑改进；
3. 默认文本国际化；
4. 文档改写，适配 dumi；
5. 测试补全

## Drawbacks
暂无。

## Bonus
1. 交错样式
2. 水印防删除功能

**Comment by AH-dark** [comment]
What is the significance of this watermark? Doesn’t deleting the `<canvas>` component mean that there is nothing? 🤨

**Comment by JarvisArt** [comment]
可以支持下下面的两个功能：
1.防删除水印dom
2.水印支持像下图这种交错显示，挺多水印是这种的
![image](https://user-images.githubusercontent.com/38420763/203471576-2a117dca-c22e-40a6-a304-36c7d4775e6c.png)

**Comment by chinyifei** [comment]
这个组件如果是之前antd4的版本，可以单独引入这个WaterMark 组件吗？如果不可以的话。4.20升级5.20会有冲突吗？

**Reply by JarvisArt** [reply]
https://www.npmjs.com/package/antd-watermark 可以先用这个包

**Comment by yuguaa** [comment]
https://github.com/yuguaa/watermark       整成了js版本，美滋滋

**Comment by JSchmidt63** [comment]
seen in: https://pro-components.antdigital.dev/en-US/components/water-mark
